### PR TITLE
Thumbnail allow upscaling

### LINKED ISSUE #28: Image::thumbnail with Upscaling?
### What steps will reproduce the problem?
Use a Image smaller than 120x120 and create a Thumbnail with Box Option

Image::thumbnail('@webroot/img/test-image.jpg', 120, 120, ImageInterface::THUMBNAIL_INSET)
->save(Yii::getAlias('@runtime/thumb-test-image.jpg'), ['quality' => 50]);

### What's expected?
The smaller Image should be placed inside the Box 120x120

### What do you get instead?
A Thumbnail with the original size. Maybe this is intened, but if you specify the size, you expect the Image to be upscaled

### Additional info

| Q                | A
| ---------------- | ---
| Yii vesion       | 2.0.10
| Image Version | 2.1.0
| PHP version      | 5.6.27
| Operating system | Mac/Linux

### PR DESCRIPTION
**Pull request by Renkas**
Added `$allowUpscaling` argument to `Image::thumbnail()` so it would be possible to always have thumbnail with fixed dimensions.

Fixes #28

### CODE REVIEW

File: BaseImage.php [hunk 1]
```diff
@@ -212,9 +212,12 @@ public static function autorotate($image, $color = '000000')
      * @param int $width the width in pixels to create the thumbnail
      * @param int $height the height in pixels to create the thumbnail
      * @param string $mode mode of resizing original image to use in case both width and height specified
+     * @param bool $allowUpscaling should the image be upscaled if needed
      * @return ImageInterface
+     *
+     * @since 2.1.1 Added the $allowUpscaling argument.
      */
-    public static function thumbnail($image, $width, $height, $mode = ManipulatorInterface::THUMBNAIL_OUTBOUND)
+    public static function thumbnail($image, $width, $height, $mode = ManipulatorInterface::THUMBNAIL_OUTBOUND, $allowUpscaling = false)
```
Comment (samdark) [inline BaseImage.php]: Since method signature was changed, it's better to release this in 2.2.0.

Comment (Renkas) [inline BaseImage.php]: changes 2.1.1 -> 2.2.0 for all since tags and in changelog as it seems 2.1.1 release will never happen anyway

**Comment by rob006** [comment]
We could add another variations for `$mode` - then we could avoid BC break and keep method signature simpler. 

It also does not cover all usecases - in one project I need images with fixed dimensions, but I don't want upscaling, so I add blank borders around image. But in some cases I don't need fixed dimensions, but I need fixed proportions. So actually there is at least 4 ways to generate thumbnail (assume that we have image with dimensions 70x50 and want to generate thumbnail with dimensions 100x100 and inset mode):
1. does not change image: want 100x100, got 70x50 (current implementation).
2. upscaling: 100x100 with upscaling and blank borders in one dimension (this PR)
3. fixed dimensions, no upscaling: 100x100 with blank borders in both dimensions
4. fixed proportions, no upscaling: 70x70 with blank borders in one dimension.

**Comment by Renkas** [comment]
@rob006 which part of this pull request breaks BC?

Your four points are separate issue. And AFAIK i understand what you need 1-3 is already implemented.

**Comment by rob006** [comment]
> which part of this pull request breaks BC?

It changes method signature, so it will break my app (I already override `Image::thumbnail()` and I will have incompatible signature).

> Your four points are separate issue. 

And how you want to solve this? By adding 2 more params for this method? It just need to be resolved smarter than just adding another boolean param.

**Comment by Renkas** [comment]
I'm sure no one will have nothing against a better solution if you have one in mind.

**Comment by rob006** [comment]
I already proposed solution - use `$mode` param and create modes like `THUMBNAIL_OUTBOUND_WITH_UPSCALING`. No need to change method signature and you can easily add another modes in the future.

**Comment by Renkas** [comment]
Will this be merged or what will happen with the pull request?

@rob006 has other idea how to solve this but has not given the solution/pull request.

I'm happy with this implementation and am forced to use my own fork until the functionality gets to official extension. Be it then my solution or rob's idea.

File: BaseImage.php [hunk 2]
```diff
@@ -415,7 +422,7 @@ protected static function getThumbnailBox(BoxInterface $sourceBox, $width, $heig
      * @param bool $keepAspectRatio should we keep aspect ratio even if both with and height are set
      * @return BoxInterface new image box
      *
-     * @since 2.1.1
+     * @since 2.2.0
```
Comment (Renkas) [inline BaseImage.php]: no. There's no 2.1.1 release

File: BaseImage.php [hunk 3]
```diff
@@ -276,7 +283,7 @@ public static function thumbnail($image, $width, $height, $mode = ManipulatorInt
      * @param bool $allowUpscaling should the image be upscaled if needed
      * @return ImageInterface
      *
-     * @since 2.1.1
+     * @since 2.2.0
```
Comment (Renkas) [inline BaseImage.php]: no. There's no 2.1.1 release

Comment (samdark) [inline BaseImage.php]: Right.

File: BaseImage.php (hunk 2 above)
Comment (samdark) [inline BaseImage.php]: Right.

**Comment by samdark** [comment]
I like what @rob006 proposed.

**Comment by samdark** [comment]
See https://github.com/yiisoft/yii2-imagine/issues/28

**Pull request closed without merging**